### PR TITLE
feat: add Linux aarch64/arm64 support for bioconda-utils

### PR DESCRIPTION
**Pull request by Yikun**
This PR try to add Linux aarch64/arm64 support for bioconda-utils:
- Support `BASE_IMAGE` build args in `Dockerfile`: the devs can specify `BASE_IMAGE` to switch base image.
- `bioconda_utils/bioconda_utils-requirements.txt`: Upgrade packages version to support Linux aarch64.
- Add `--docker-base-image` image to help users specify docker base image.
- Add Linux aarch64 image build to `release-please.yml` and `build-image.yml` workflow
- `bioconda_utils/docker_utils.py`: create conda build subdir according to arch.

After this PR, users can build the bioconda package by using same cmd with x86 by specifing `--docker-base-image` to build the  package. 

Note that the `mulled-test` are still not supported, there are a circle dependency between `bioconda-containers` and bioconda-utils, we'd better to support bioconda-utils aarch64 image first.

Related: https://github.com/bioconda/bioconda-utils/issues/706

Test step:
```bash

$ uname -m
aarch64

# Activate a conda env
conda create -n biotest
conda activate biotest

git clone https://github.com/Yikun/bioconda-utils.git
cd bioconda-utils
git checkout aarch64

# Build docker image with embed involucro for bioconda
docker build -t ghcr.io/yikun/bioconda-utils-build-env-cos7-aarch64 -f Dockerfile .

# Install all bioconda-utils
conda install --file bioconda_utils/bioconda_utils-requirements.txt -c conda-forge -c bioconda

# Install the bioconda-utils with this PR
pip install -e .

# Test
bioconda-utils -h

# Build a package to test
cd bioconda-recipes

bioconda-utils build --docker --packages bamstats --docker-base-image ghcr.io/yikun/bioconda-utils-build-env-cos7-aarch64 --force
```

### CODE REVIEW

File: Dockerfile.aarch64
Comment (martin-g) [inline Dockerfile.aarch64:6]:
```suggestion
FROM $BASE_IMAGE as base
```

File: Dockerfile.aarch64
Comment (martin-g) [inline Dockerfile.aarch64:8]: The comment says `from our base image` but it uses totally different image - `bioconda/base-glibc-busybox-bash`.
Is it possible to use ` $BASE_IMAGE` instead ?

Comment (Yikun) [reply Dockerfile.aarch64:8]: This base image is not `${BASE_IMAGE}`, I'm not sure this should be address or not, this is copy from x86 dockerfile. So if we change this, we can fix them in a separate PR.

**Comment by Yikun** [comment]
Test passed in my local fork: https://github.com/Yikun/bioconda-utils/pull/2

**Comment by bgruening** [comment]
This one needs a rebase.

**Comment by Yikun** [comment]
A question about `divergent_build`: 

https://github.com/bioconda/bioconda-utils/blob/1923d24c4f3cd38740ecfbf240b92d5eb1432e09/bioconda_utils/utils.py#L1136-L1140

With this PR, all aarch64 pkg **without verison bump** (even with the `--force`) will rasie the exception here, and finally failed on:

https://github.com/bioconda/bioconda-utils/blob/1923d24c4f3cd38740ecfbf240b92d5eb1432e09/bioconda_utils/build.py#L324-L327

I'm not sure what's the meaning this exception? What's the meaning of `divergent build strings`?

cc @bgruening @epruesse Is this behavior normal? Would you give some idea on here? 

(I also dump the values compared to x86 to see how divergent_builds work, I still can't understand well)
https://github.com/bioconda/bioconda-utils/blob/1923d24c4f3cd38740ecfbf240b92d5eb1432e09/bioconda_utils/utils.py#L1119
```
# existing_pkg_builds - set(build_meta.keys())
# aarch64: {Pandas(subdir='linux-64', build='he881be0_4')} - {('linux-aarch64', 'h8865c2f_4')}
# x86_64: {Pandas(subdir='linux-64', build='he881be0_4')} - {('linux-64', 'he881be0_4')}
```

**Comment by bgruening** [comment]
> I think I understand the meaning of divergent_build, it's for detecting changes on build but no version bump for specific arch/noarch. So I beliebe we need changes like: https://github.com/bioconda/bioconda-utils/commit/9e7b0fd95736d921ebabb0c982117e42997ebfd0#diff-79a86610eedbe72fc010927f2fec251188ef01ec79e21f8ddef67d912cda4295 .

Yes, see in matrix :)

What is your deployment strategy for the entire repo? Should we have an initial allow list and migrate single recipes one by one?

**Comment by Yikun** [comment]
> What is your deployment strategy for the entire repo? Should we have an initial allow list and migrate single recipes one by one?

@bgruening  I thought the strategy might be:

Step 1: Support bioconda-utils on Linux aarch64. (This PR)
- `pip install bioconda-utils`: User can install the `bioconda-utils` package directly in Linux aarch64
- `bioconda-utils build --docker --packages XXX --docker-base-image bioconda-utils-build-env-cos7-aarch64:latest --force`: User can build the bioconda package

Step 2: Support bioconda-utils mulled test on Linux aarch64 (depends on https://github.com/bioconda/bioconda-containers/pull/55)
- `bioconda-utils build --docker --mulled-test --packages XXX --docker-base-image bioconda-utils-build-env-cos7-aarch64:latest --force`: User can build and test bioconda package.

(After step1 / step2 completed, users can build their wanted packages in a local aarch64 env)

Step 3: Run bulk test on Linux aarch64 and get build/test results
- `bioconda-utils build recipes config.yml --worker-offset 0 --n-workers N --docker --mulled-test --force`: build and run all tests and find all failure packages. Then we will have an `initial allow list` candidate package. And in this step we might get a page to show which packages build passed or failure, the potential style might be something like what we done in [bioconductor report](https://yikun.github.io/bioconductor-0301/report/long-report.html) or conda-forge team [migration page](https://conda-forge.org/status/#aarch64andppc64leaddition).
- Fix the package one by one, and the results page will generated peridically to guide us which packages should be fixed.

Step 4: Upload all passed package and enable Linux aarch64 CI.
- This step required the Linux aarch64 resource (need resource donation) or qemu on github action (might slow).
- We could select a few and then dozens of packages as initial allow list to validate the linux aarch64 build/test/upload workflow
- After complete the initial validation, we can change `allow list` to `block list`, or `full list` and enable Linux aarch64 CI.

This is just an initial idea, there is no doubt that the bioconda team has a more professional perspective, so we'd also like to hear your suggestions, any thoughts?

**Comment by Yikun** [comment]
I pick two packages randomly to test the build / test / push step:

1. bamstats
    - `bioconda-utils build --docker --mulled-test --packages bamstats --docker-base-image ghcr.io/yikun/bioconda-utils-build-env-cos7-aarch64 --mulled-conda-image ghcr.io/martin-g/create-env-aarch64 --anaconda-upload --force`
    - https://anaconda.org/yikunkero/bamstats

2. py2bit
    - `bioconda-utils build --docker --mulled-test --packages bioconda-utils --docker-base-image ghcr.io/yikun/bioconda-utils-build-env-cos7-aarch64 --mulled-conda-image ghcr.io/martin-g/create-env-aarch64 --anaconda-upload --force`
    - https://anaconda.org/yikunkero/py2bit

All is well! 

Feel free to tell me if any other preworks should be done before merged.

**Comment by Yikun** [comment]
@bgruening is there any way to run all package build/test without failfast but also consider dependency?

```
# This will break after first package failed
bioconda-utils build recipes config.yml --worker-offset 0 --n-workers 1 --docker biocontainers --force
```

I hope there is a way like:

```
ll /root/bio/bioconda-recipes | xargs -n1 bioconda-utils build --docker --mulled-test --force --packages
```

but also consider the dependency.

**Comment by bgruening** [comment]
@Yikun yes this should be possible. We are doing this on the bulk branch. I think if you try your rebuild on this branch it should work. Background is that we use this branch to bulk-rebuild the entire tree against new version-pins and 100 of package-updates at once.

**Comment by martin-g** [comment]
What command one should use to build multiple packages ? In parallel or sequentially, but without stopping on the first failed package.

I tried
```
bioconda-utils build recipes config.yml --worker-offset 0 --n-workers 5000 --docker --mulled-test --anaconda-upload --docker-base-image ghcr.io/yikun/bioconda-utils-build-env-cos7-aarch64 --mulled-conda-image ghcr.io/martin-g/create-env-aarch64
```
on both `master` and `bulk` branches but it always fails on:
```
13:03:48 BIOCONDA INFO (OUT)      Running `rustc --crate-name rust_htslib --edition=2018 /opt/conda/conda-bld/10x_bamtofastq_1681822376637/_build_env/.cargo/registry/src/github.com-1ecc6299db9ec823/rust-htslib-0.38.2/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C linker-plugin-lto -C debuginfo=1 --cfg 'feature="bzip2"' --cfg 'feature="libdeflate"' --cfg 'feature="lzma"' -C metadata=21249687d7a2ff42 -C extra-filename=-21249687d7a2ff42 --out-dir /opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps -C linker=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/_build_env/bin/aarch64-conda-linux-gnu-cc -L dependency=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps --extern bio_types=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/libbio_types-b44ae3b4e5106054.rmeta --extern byteorder=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/libbyteorder-3fa2d67d6d18ade5.rmeta --extern custom_derive=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/libcustom_derive-ffaeddde1e18b352.rmeta --extern derive_new=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/libderive_new-ad983a2555eabb00.so --extern hts_sys=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/libhts_sys-99c9abb0944d285f.rmeta --extern ieee754=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/libieee754-075bde6610681de0.rmeta --extern lazy_static=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/liblazy_static-1464c69ac8c9cd33.rmeta --extern libc=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/liblibc-9ed5c1a1ce845d01.rmeta --extern linear_map=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/liblinear_map-e477efefa8fc49d6.rmeta --extern newtype_derive=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/libnewtype_derive-d631dfb49616692a.rmeta --extern regex=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/libregex-dffaa8dd80e1dd48.rmeta --extern thiserror=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/libthiserror-054997a83c8358b5.rmeta --extern url=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/deps/liburl-9ab6a5e77ff5d9aa.rmeta --cap-lints allow -L native=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/build/hts-sys-3bb7368342e96f5e/out -L native=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/build/bzip2-sys-86a1869615adb9cc/out/lib -L native=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/build/libdeflate-sys-169cab9152541b0c/out/lib -L native=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/build/libz-sys-869a20e7c98a6611/out/lib -L native=/opt/conda/conda-bld/10x_bamtofastq_1681822376637/work/target/release/build/lzma-sys-c093ea58537dae4b/out`
13:03:49 BIOCONDA INFO (OUT) error[E0308]: mismatched types
13:03:49 BIOCONDA INFO (OUT)    --> /opt/conda/conda-bld/10x_bamtofastq_1681822376637/_build_env/.cargo/registry/src/github.com-1ecc6299db9ec823/rust-htslib-0.38.2/src/bam/record.rs:152:16
13:03:49 BIOCONDA INFO (OUT)     |
13:03:49 BIOCONDA INFO (OUT) 152 |             s: sam_copy.as_ptr() as *mut i8,
13:03:49 BIOCONDA INFO (OUT)     |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u8`, found `i8`
13:03:49 BIOCONDA INFO (OUT)     |
13:03:49 BIOCONDA INFO (OUT)     = note: expected raw pointer `*mut u8`
13:03:49 BIOCONDA INFO (OUT)                found raw pointer `*mut i8`
13:03:49 BIOCONDA INFO (OUT) 
```
for `10x_bamtofastq` package.
I've requested a new release of this package with a fix - https://github.com/10XGenomics/bamtofastq/issues/136. 
What command I could use until then ?

**Comment by Yikun** [comment]
> What command one should use to build multiple packages ? In parallel or sequentially, but without stopping on the first failed package.

I think it should be:
```
ls /root/bio/bioconda-recipes/recipes/ | xargs -n1 bioconda-utils build --docker --mulled-test --docker-base-image ghcr.io/yikun/bioconda-utils-build-env-cos7-aarch64 --mulled-conda-image ghcr.io/martin-g/create-env-aarch64 --anaconda-upload --packages 
```

File: .github/workflows/release-please.yml
Comment (johanneskoester) [inline .github/workflows/release-please.yml:32]: Why did you remove this?

**Review by johanneskoester — CHANGES_REQUESTED** [review]

File: .github/workflows/release-please.yml
Comment (Yikun) [reply .github/workflows/release-please.yml:32]: Oh, good catch. It should bad copy paste from `build-image.yml`.

Resolved!

File: .github/workflows/release-please.yml
Comment (Yikun) [reply .github/workflows/release-please.yml:15]: @johanneskoester Would you mind taking a look? Maybe we need some configure on Quay bioconda repo?

**Comment by milot-mirdita** [comment]
Does this work already?

I would like to make my recipes available for linux arm64. They would need a different `cmake` flag thought.
Would `uname -m` be the recommended mechanism to set arm64 specific flags?

**Comment by Yikun** [comment]
Now you can build the aarch64 package in your linux arm64 env locally by below cmd:
```
bioconda-utils build --docker --packages YOUR_PACKAGE_NAME --docker-base-image ghcr.io/yikun/bioconda-utils-build-env-cos7-aarch64 --force
```

We still need to support [1][2] before pulish the bioconda packages.

[1] Support mulled test: https://github.com/bioconda/bioconda-containers/pull/55
[2] Offcial publish: https://github.com/bioconda/bioconda-recipes/pull/40550

> Would uname -m be the recommended mechanism to set arm64 specific flags?

`uname -m` is a good way to set linux aarch64 flags. I guess you might also interested on [config.guess](https://git.savannah.gnu.org/cgit/config.git/plain/config.guess).

**Comment by milot-mirdita** [comment]
Thanks for the update! I'll wait until the two steps are merged/completed.

File: .github/workflows/release-please.yml
Comment (Yikun) [reply .github/workflows/release-please.yml:15]: also cc @bgruening

https://github.com/bioconda/bioconda-utils/actions/runs/5471986195/jobs/9963670995 Would you mind taking a look?

Comment (martin-g) [reply .github/workflows/release-please.yml:15]: Friendly ping!
Could someone please push the new Docker image for aarch64 ? Thank you!

**Pull request merged**
